File: hive/src/main/java/com/dbx/agent/hive/HiveAgent.java
package com.dbx.agent.hive;

import com.dbx.agent.AbstractJdbcAgent;
import com.dbx.agent.ColumnInfo;
import com.dbx.agent.ConnectParams;
import com.dbx.agent.DatabaseInfo;
import com.dbx.agent.ForeignKeyInfo;
import com.dbx.agent.IndexInfo;
import com.dbx.agent.JdbcIdentifiers;
import com.dbx.agent.JsonRpcServer;
import com.dbx.agent.TableInfo;
import com.dbx.agent.TriggerInfo;
import java.sql.ResultSet;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

public final class HiveAgent extends AbstractJdbcAgent {

    @Override
    protected String driverClass() {
        return "org.apache.hive.jdbc.HiveDriver";
    }

    @Override
    protected String buildJdbcUrl(ConnectParams params) {
        return buildUrl(params);
    }

    @Override
    public List<DatabaseInfo> listDatabases() {
        return unchecked(() -> {
            List<DatabaseInfo> result = new ArrayList<>();
            try (java.sql.Statement stmt = requireConnected().createStatement();
                 ResultSet rs = stmt.executeQuery("SHOW DATABASES")) {
                while (rs.next()) {
                    result.add(new DatabaseInfo(rs.getString(1)));
                }
            }
            result.sort(Comparator.comparing(DatabaseInfo::getName));
            return result;
        });
    }

    @Override
    public List<String> listSchemas() {
        List<String> result = new ArrayList<>();
        for (DatabaseInfo database : listDatabases()) {
            result.add(database.getName());
        }
        return result;
    }

    @Override
    public List<TableInfo> listTables(String schema) {
        return unchecked(() -> {
            List<TableInfo> result = new ArrayList<>();
            useSchema(schema);
            try (java.sql.Statement stmt = requireConnected().createStatement();
                 ResultSet rs = stmt.executeQuery("SHOW TABLES")) {
                while (rs.next()) {
                    result.add(new TableInfo(rs.getString(1), "TABLE", null));
                }
            }
            result.sort(Comparator.comparing(TableInfo::getName));
            return result;
        });
    }

    @Override
    public List<ColumnInfo> getColumns(String schema, String table) {
        return unchecked(() -> {
            List<ColumnInfo> result = new ArrayList<>();
            useSchema(schema);
            try (java.sql.Statement stmt = requireConnected().createStatement();
                 ResultSet rs = stmt.executeQuery("DESCRIBE " + JdbcIdentifiers.INSTANCE.backtick(table))) {
                while (rs.next()) {
                    String colName = trimToNull(rs.getString(1));
                    if (colName == null || colName.startsWith("#")) {
                        continue;
                    }
                    result.add(new ColumnInfo(
                        colName,
                        trimToEmpty(rs.getString(2)),
                        true,
                        null,
                        false,
                        null,
                        optionalComment(rs),
                        null,
                        null,
                        null
                    ));
                }
            }
            return result;
        });
    }

    @Override
    public List<IndexInfo> listIndexes(String schema, String table) {
        return Collections.emptyList();
    }

    @Override
    public List<ForeignKeyInfo> listForeignKeys(String schema, String table) {
        return Collections.emptyList();
    }

    @Override
    public List<TriggerInfo> listTriggers(String schema, String table) {
        return Collections.emptyList();
    }

    @Override
    public String setSchemaSQL(String schema) {
        return "USE " + JdbcIdentifiers.INSTANCE.backtick(schema);
    }

    @Override
    protected Object resultValue(ResultSet rs, int index, int sqlType) {
        return unchecked(() -> {
            Object value = rs.getObject(index);
            return rs.wasNull() ? null : value == null ? null : value.toString();
        });
    }

    private void useSchema(String schema) throws Exception {
        try (java.sql.Statement stmt = requireConnected().createStatement()) {
            stmt.execute(setSchemaSQL(schema));
        }
    }

    private static String buildUrl(ConnectParams params) {
        return "jdbc:hive2://" + params.getHost() + ":" + params.getPort() + "/" + params.getDatabase();
    }

    private static String trimToNull(String value) {
        if (value == null) {
            return null;
        }
        String trimmed = value.trim();
        return trimmed.isEmpty() ? null : trimmed;
    }

    private static String trimToEmpty(String value) {
        return value == null ? "" : value.trim();
    }

    private static String optionalComment(ResultSet rs) {
        return unchecked(() -> {
            try {
                String comment = trimToNull(rs.getString(3));
                return comment;
            } catch (Exception e) {
                return null;
            }
        });
    }

    public static void main(String[] args) {
        new JsonRpcServer(new HiveAgent()).run();
    }
}

File: hive/src/main/kotlin/com/dbx/agent/hive/HiveAgent.kt
package com.dbx.agent.hive

import com.dbx.agent.*
import java.sql.Connection
import java.sql.DriverManager

class HiveAgent : DatabaseAgent {

    private var connection: Connection? = null

    override fun getConnection(): Connection? = connection

    override fun connect(params: ConnectParams) {
        Class.forName("org.apache.hive.jdbc.HiveDriver")
        val url = "jdbc:hive2://${params.host}:${params.port}/${params.database}"
        connection = DriverManager.getConnection(url, params.username, params.password)
    }

    override fun testConnection(params: ConnectParams): Boolean {
        Class.forName("org.apache.hive.jdbc.HiveDriver")
        val url = "jdbc:hive2://${params.host}:${params.port}/${params.database}"
        DriverManager.getConnection(url, params.username, params.password).use { conn ->
            return conn.isValid(5)
        }
    }

    override fun listDatabases(): List<DatabaseInfo> {
        val conn = requireConnection()
        conn.createStatement().use { stmt ->
            stmt.executeQuery("SHOW DATABASES").use { rs ->
                return buildList {
                    while (rs.next()) {
                        add(DatabaseInfo(rs.getString(1)))
                    }
                }.sortedBy { it.name }
            }
        }
    }

    override fun listSchemas(): List<String> {
        return listDatabases().map { it.name }
    }

    override fun listTables(schema: String): List<TableInfo> {
        val conn = requireConnection()
        // Switch to the target database first
        conn.createStatement().use { stmt ->
            stmt.execute("USE ${JdbcIdentifiers.backtick(schema)}")
        }
        conn.createStatement().use { stmt ->
            stmt.executeQuery("SHOW TABLES").use { rs ->
                return buildList {
                    while (rs.next()) {
                        add(TableInfo(
                            name = rs.getString(1),
                            table_type = "TABLE"
                        ))
                    }
                }.sortedBy { it.name }
            }
        }
    }

    override fun getColumns(schema: String, table: String): List<ColumnInfo> {
        val conn = requireConnection()
        // Switch to the target database first
        conn.createStatement().use { stmt ->
            stmt.execute("USE ${JdbcIdentifiers.backtick(schema)}")
        }
        conn.createStatement().use { stmt ->
            stmt.executeQuery("DESCRIBE ${JdbcIdentifiers.backtick(table)}").use { rs ->
                return buildList {
                    while (rs.next()) {
                        val colName = rs.getString(1)?.trim() ?: continue
                        // DESCRIBE output may include partition info headers (empty lines or "# col_name")
                        if (colName.isEmpty() || colName.startsWith("#")) continue
                        val dataType = rs.getString(2)?.trim() ?: ""
                        val comment = try { rs.getString(3)?.trim()?.ifEmpty { null } } catch (_: Exception) { null }

                        add(ColumnInfo(
                            name = colName,
                            data_type = dataType,
                            is_nullable = true, // Hive columns are nullable by default
                            column_default = null,
                            is_primary_key = false, // Hive does not have traditional primary keys
                            extra = null,
                            comment = comment,
                            numeric_precision = null,
                            numeric_scale = null,
                            character_maximum_length = null
                        ))
                    }
                }
            }
        }
    }

    override fun listIndexes(schema: String, table: String): List<IndexInfo> {
        // Hive does not have traditional indexes
        return emptyList()
    }

    override fun listForeignKeys(schema: String, table: String): List<ForeignKeyInfo> {
        // Hive does not have foreign keys
        return emptyList()
    }

    override fun listTriggers(schema: String, table: String): List<TriggerInfo> {
        // Hive does not have triggers
        return emptyList()
    }

    override fun executeQuery(sql: String, schema: String?, options: ExecuteQueryOptions): QueryResult {
        return JdbcExecutor.execute(requireConnection(), sql, schema, ::setSchemaSQL, maxRows = options.maxRows, fetchSize = options.fetchSize, valueReader = ::stringResultValue)
    }

    override fun setSchemaSQL(schema: String): String = "USE ${JdbcIdentifiers.backtick(schema)}"

    override fun disconnect() {
        connection?.close()
        connection = null
    }

    private fun requireConnection(): Connection {
        return connection ?: throw IllegalStateException("Not connected")
    }

    private fun stringResultValue(rs: java.sql.ResultSet, index: Int, sqlType: Int): Any? {
        val value = rs.getObject(index)
        return if (rs.wasNull()) null else value?.toString()
    }
}

fun main() {
    JsonRpcServer(HiveAgent()).run()
}
